### PR TITLE
semaphore fix

### PR DESCRIPTION
**Pull request by brthor**
@eerhardt @ellismg @piotrpMSFT

### CODE REVIEW

File: build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs [hunk 1]
```diff
@@ -103,12 +105,30 @@ public IEnumerable<string> ListBlobs(string virtualDirectory)
...
-            CloudBlockBlob cloudBlob = _blobContainer.GetBlockBlobReference(blob);
-            System.Threading.Tasks.Task<string> task = cloudBlob.AcquireLeaseAsync(TimeSpan.FromMinutes(1), null);
-            task.Wait(); 
-            return task.Result;
+            StopWatch stopWatch = new StopWatch();
+            stopWatch.Start();
+
+            // This will throw an exception with HTTP code 409 when we cannot acquire the lease
+            // But we should block until we can get this lease, with a timeout (maxWaitSeconds)
+            while (stopWatch.ElapsedMilliseconds < maxWaitMilliseconds)
```
Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: does this build?

Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: parameter needs to be renamed.

File: build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs [hunk 2]
```diff
@@ -103,12 +105,30 @@ public IEnumerable<string> ListBlobs(string virtualDirectory)
             return blobFiles.Results.Select(bf => bf.Uri.PathAndQuery);
         }
 
-        public string AcquireLeaseOnBlob(string blob)
+        public string AcquireLeaseOnBlob(string blob, int maxWaitSeconds=120000, int delayMilliseconds=500)
```
Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: Let's make these TimeSpans instead. They read a lot better.

File: build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs (hunk 1 above)
Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: nope on Sublime, fixing it locally

File: build_projects/dotnet-host-build/PublishTargets.cs [hunk 3]
```diff
@@ -70,9 +73,10 @@ public static BuildTargetResult FinalizeBuild(BuildTargetContext c)
             if (CheckIfAllBuildsHavePublished())
             {
                 string targetContainer = $"{Channel}/Binaries/Latest/";
-                string targetVersionFile = $"{targetContainer}{SharedFrameworkNugetVersion}";
+                string targetVersionFile = $"{targetContainer}{BranchName}-{CommitCount}";
```
Comment (piotrpMSFT) [inline build_projects/dotnet-host-build/PublishTargets.cs]: per offline, let's just replace this with SHA of the commit. It encompasses CommitCount and disambiguates across branches.

Comment (brthor) [inline build_projects/dotnet-host-build/PublishTargets.cs]: done.

File: build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs (hunk 2 above)
Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: done.

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: can't use them, must be compile-time constant

Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: TimeSpan.Zero and then set them in code?

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: don't you think it defeats the purpose of a default?

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: I'll do null, then in code do `maxWait = maxWait ?? TimeSpan....`

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: dude I have never had so much trouble with default values. Timespans aren't going to work here

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: `??` doesn't work on TimeSpans, Timespans can't be null, TimeSpan.Zero is not a compile time constant

Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: `TimeSpan?`  -- nullable timespan

Comment (eerhardt) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]:
```
 public string AcquireLeaseOnBlob(string blob, TimeSpan? maxWait=null, TimeSpan? delay=null)
 {
   maxWait = maxWait ?? new TimeSpan(blah);

.....

   maxWait.Value gets you the TimeSpan
 }
```

Comment (brthor) [inline build_projects/shared-build-targets-utils/Publishing/AzurePublisher.cs]: one last try

File: build_projects/shared-build-targets-utils/project.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
     "Microsoft.CSharp": "4.0.1",
     "System.Dynamic.Runtime": "4.0.11",
     "System.Reflection.Metadata": "1.3.0",
+    "System.Runtime.Extensions":"4.1.0",
```
Comment (eerhardt) [inline build_projects/shared-build-targets-utils/project.json]: Weird that this isn't in NetStandard.Library...

**Comment by eerhardt** [comment]
:shipit:

File: build_projects/shared-build-targets-utils/project.json (hunk 4 above)
Comment (brthor) [inline build_projects/shared-build-targets-utils/project.json]: I thought that too, sure there's reasoning behind it though

Comment (weshaggard) [inline build_projects/shared-build-targets-utils/project.json]: That is in NETStandard.Library. https://github.com/dotnet/corefx/blob/master/pkg/NETStandard.Library/NETStandard.Library.packages.targets#L30 

What issue were you hitting?

**Pull request merged**